test/core/base/cache.js

```diff
@@ -1,32 +1,173 @@
+      function fn() {
+        axe._cache.set('foo', 'bar');
+      }
+      assert.doesNotThrow(fn);
+    });
+
+    // undefined is a valid result type of a potentially expensive operation in axe-core
+    it('should set `undefined` without error', function () {
+      axe._cache.set('foo', undefined);
+
+      function fn() {
+        assert.isUndefined(axe._cache.get('foo'));
+      }
+      assert.doesNotThrow(fn);
+    });
+
+    it('should throw TypeError for non-string keys', function () {
+      assert.throws(function () {
+        axe._cache.set(1, 'bar');
+      }, TypeError);
+      assert.throws(function () {
+        axe._cache.set({}, 'bar');
+      }, TypeError);
+      assert.throws(function () {
+        axe._cache.set(null, 'bar');
+      }, TypeError);
+      assert.throws(function () {
+        axe._cache.set(function () {
+          return;
+        }, 'bar');
+      }, TypeError);
+    });
 
-  it('should not throw for non-string keys', function() {
-    function fn() {
-      axe._cache.set(1, 'bar');
-      axe._cache.set({}, 'bar');
-      axe._cache.set(null, 'bar');
-    }
-    assert.doesNotThrow(fn);
+    it('should throw TypeError for empty string keys', function () {
+      assert.throws(function () {
+        axe._cache.set('', 'bar');
+      }, TypeError);
+    });
   });
 
-  it('should get an item from the cache', function() {
-    axe._cache.set('foo', 'bar');
-    var value = axe._cache.get('foo');
-    assert.equal(value, 'bar');
+  describe('get', function () {
+    it('should get an item from the cache', function () {
+      axe._cache.set('foo', 'bar');
+      var value = axe._cache.get('foo');
+      assert.equal(value, 'bar');
+    });
+
+    it('should return `undefined` for key lookup miss', function () {
+      assert.isUndefined(axe._cache.get('foo'));
+    });
+
+    describe('with creator function', function () {
+      it('should set and return primitive types', function () {
+        assert.equal(
+          axe._cache.get('integer', function () {
+            return 1;
+          }),
+          1
+        );
+        assert.equal(
+          axe._cache.get('float', function () {
+            return 1.1;
+          }),
+          1.1
+        );
+        assert.equal(
+          axe._cache.get('string', function () {
+            return 'foo';
+          }),
+          'foo'
+        );
+        assert.isTrue(
+          axe._cache.get('boolean', function () {
+            return true;
+          })
+        );
+        assert.isNull(
+          axe._cache.get('null', function () {
+            return null;
+          })
+        );
+
+        var obj = { foo: 'bar' };
+        assert.equal(
+          axe._cache.get('object', function () {
+            return obj;
+          }),
+          obj
+        );
+
+        assert.sameOrderedMembers(
+          axe._cache.get('array', function () {
+            return [1, 2, 3];
+          }),
+          [1, 2, 3]
```

Comment:
You'll want to check that it returns the same array like you did for the obj

```suggestion
        var array = [1, 2, 3];
        assert.equal(
          axe._cache.get('array', function () {
            return array;
          }),
          array
```